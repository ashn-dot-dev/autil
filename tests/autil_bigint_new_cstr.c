#define AUTIL_IMPLEMENTATION
#include "../autil.h"
#include "test.h"

static void
test_valid(char const* input, char const* fmt)
{
    assert(input != NULL);
    assert(fmt != NULL);

    struct autil_bigint* const num = autil_bigint_new_cstr(input);
    ASSERT(num != NULL);

    char* const cstr = autil_bigint_to_new_cstr(num, fmt);
    ASSERT(cstr != NULL);

    ASSERT((strlen(cstr) <= 36) && "NEED TO INCREASE PRINTF FORMAT LENGTH");
    printf("%-36s | %s\n", input, cstr);

    autil_bigint_del(num);
    autil_xalloc(cstr, AUTIL_XALLOC_FREE);
}

int
main(void)
{
    printf("%-36s | %s\n", "INPUT", "OUTPUT");

    test_valid("0", "#+d");
    test_valid("+0", "#+d");
    test_valid("-0", "#+d");

    test_valid("1", "#+d");
    test_valid("+1", "#+d");
    test_valid("-1", "#+d");
    test_valid("0b1", "#+b");
    test_valid("+0b1", "#+b");
    test_valid("-0b1", "#+b");
    test_valid("0x1", "#+x");
    test_valid("+0x1", "#+x");
    test_valid("-0x1", "#+x");

    test_valid("0b111100001100110010101010", "#+b");
    test_valid("+0b111100001100110010101010", "#+b");
    test_valid("-0b111100001100110010101010", "#+b");

    test_valid("0xDEADBEEF", "#+x");
    test_valid("+0xDEADBEEF", "#+x");
    test_valid("-0xDEADBEEF", "#+x");
    test_valid("0xcafed00d", "#+x");
    test_valid("0xcafebabe", "#+x");
    test_valid("0xbeef0000", "#+x");
    test_valid("0xBEEFBEEFBEEFBEEFBEEFBEEFBEEFBEEF", "#+x");

    test_valid("00", "#+d");
    test_valid("00000000000000000000000000000000", "#+d");

    EXPECT(NULL == autil_bigint_new_cstr(""));
    EXPECT(NULL == autil_bigint_new_cstr("+"));
    EXPECT(NULL == autil_bigint_new_cstr("-"));
    EXPECT(NULL == autil_bigint_new_cstr(" "));
    EXPECT(NULL == autil_bigint_new_cstr(" 1"));
    EXPECT(NULL == autil_bigint_new_cstr("1 "));
    EXPECT(NULL == autil_bigint_new_cstr(" 1 "));
    EXPECT(NULL == autil_bigint_new_cstr("1.1"));

    return EXIT_SUCCESS;
}
