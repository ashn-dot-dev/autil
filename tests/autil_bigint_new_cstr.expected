INPUT                                | OUTPUT
0                                    | 0
+0                                   | 0
-0                                   | 0
1                                    | +1
+1                                   | +1
-1                                   | -1
0b1                                  | +0b1
+0b1                                 | +0b1
-0b1                                 | -0b1
0x1                                  | +0x1
+0x1                                 | +0x1
-0x1                                 | -0x1
0b111100001100110010101010           | +0b111100001100110010101010
+0b111100001100110010101010          | +0b111100001100110010101010
-0b111100001100110010101010          | -0b111100001100110010101010
0xDEADBEEF                           | +0xdeadbeef
+0xDEADBEEF                          | +0xdeadbeef
-0xDEADBEEF                          | -0xdeadbeef
0xcafed00d                           | +0xcafed00d
0xcafebabe                           | +0xcafebabe
0xbeef0000                           | +0xbeef0000
0xBEEFBEEFBEEFBEEFBEEFBEEFBEEFBEEF   | +0xbeefbeefbeefbeefbeefbeefbeefbeef
00                                   | 0
00000000000000000000000000000000     | 0
